File: services/quant-api/app/models/review.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from sqlalchemy import Boolean, DateTime, Float, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.types import JSON

from app.db.base import Base
from app.models.data_center import TimestampMixin, utc_now


class ReviewNote(Base, TimestampMixin):
    __tablename__ = "review_notes"
    __table_args__ = (UniqueConstraint("source_type", "source_id", name="uq_review_source"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    source_type: Mapped[str] = mapped_column(String(32), index=True)
    source_id: Mapped[int | None] = mapped_column(Integer, index=True)
    symbol: Mapped[str | None] = mapped_column(String(32), index=True)
    contract: Mapped[str | None] = mapped_column(String(64), index=True)
    period: Mapped[str | None] = mapped_column(String(16), index=True)
    direction: Mapped[str | None] = mapped_column(String(16), index=True)
    strategy_name: Mapped[str | None] = mapped_column(String(64), index=True)
    strategy_version: Mapped[str | None] = mapped_column(String(32), index=True)
    open_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), index=True)
    close_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), index=True)
    open_price: Mapped[float | None] = mapped_column(Float)
    close_price: Mapped[float | None] = mapped_column(Float)
    volume: Mapped[int | None] = mapped_column(Integer)
    net_pnl: Mapped[float | None] = mapped_column(Float, index=True)
    entry_reason: Mapped[str | None] = mapped_column(Text)
    exit_reason: Mapped[str | None] = mapped_column(Text)
    market_phase: Mapped[str | None] = mapped_column(String(32), index=True)
    is_system_compliant: Mapped[bool | None] = mapped_column(Boolean, index=True)
    mistake_tags: Mapped[list[str]] = mapped_column(JSON, default=list)
    rule_tags: Mapped[list[str]] = mapped_column(JSON, default=list)
    emotion_tags: Mapped[list[str]] = mapped_column(JSON, default=list)
    lesson: Mapped[str | None] = mapped_column(Text)
    screenshot_paths: Mapped[list[str]] = mapped_column(JSON, default=list)
    kline_focus_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    kline_window_start: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    kline_window_end: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    review_score: Mapped[int | None] = mapped_column(Integer)
    ai_summary: Mapped[str | None] = mapped_column(Text)
    ai_status: Mapped[str] = mapped_column(String(32), default="reserved", index=True)
    ai_model: Mapped[str | None] = mapped_column(String(64))
    ai_generated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    extra: Mapped[dict[str, Any]] = mapped_column(JSON, default=dict)


class ReviewTag(Base, TimestampMixin):
    __tablename__ = "review_tags"
    __table_args__ = (UniqueConstraint("tag_type", "name", name="uq_review_tag_type_name"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    tag_type: Mapped[str] = mapped_column(String(32), index=True)
    name: Mapped[str] = mapped_column(String(64), index=True)
    description: Mapped[str | None] = mapped_column(Text)
    sort_order: Mapped[int] = mapped_column(Integer, default=100)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, index=True)


class ReviewAttachment(Base):
    __tablename__ = "review_attachments"

    id: Mapped[int] = mapped_column(primary_key=True)
    review_id: Mapped[int] = mapped_column(Integer, index=True)
    file_path: Mapped[str] = mapped_column(Text)
    file_type: Mapped[str | None] = mapped_column(String(32), index=True)
    title: Mapped[str | None] = mapped_column(String(128))
    meta: Mapped[dict[str, Any]] = mapped_column(JSON, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, index=True)

File: apps/quant-web/src/types/review.ts
import type { CanonicalDatasetKey, CanonicalInputIdentity } from './canonical'
import type { BarData } from './market'

export type ReviewSourceType = 'strategy_signal' | 'signal_event' | 'signal_decision' | 'manual_trade' | string

/** 单笔复盘笔记；来源仅使用独立 source_type/source_id 合同。 */
export interface ReviewNote {
  id: number
  source_type: ReviewSourceType
  source_id?: number | null
  symbol?: string | null
  contract?: string | null
  period?: string | null
  entry_interval?: string | null
  direction?: 'long' | 'short' | null
  strategy_name?: string | null
  strategy_version?: string | null
  open_time?: string | null
  close_time?: string | null
  entry_time?: string | null
  exit_time?: string | null
  open_price?: number | null
  close_price?: number | null
  volume?: number | null
  net_pnl?: number | null
  hold_bars?: number | null
  entry_reason?: string | null
  exit_reason?: string | null
  market_phase?: string | null
  is_system_compliant?: boolean | null
  mistake_tags: string[]
  setup_tags: string[]
  rule_tags: string[]
  emotion_tags: string[]
  execution_note?: string | null
  improvement_note?: string | null
  lesson?: string | null
  screenshot_path?: string | null
  screenshot_paths: string[]
  kline_focus_time?: string | null
  kline_window_start?: string | null
  kline_window_end?: string | null
  review_score?: number | null
  ai_summary?: string | null
  ai_status: string
  ai_model?: string | null
  ai_generated_at?: string | null
  extra: Record<string, unknown>
}

export interface ReviewTag {
  id: number
  tag_type: 'mistake' | 'market_phase' | 'entry_rule' | 'exit_rule' | 'emotion'
  name: string
  description?: string | null
  sort_order: number
  is_active: boolean
}

export interface ReviewUpdateRequest {
  entry_reason?: string | null
  exit_reason?: string | null
  market_phase?: string | null
  is_system_compliant?: boolean | null
  mistake_tags?: string[]
  setup_tags?: string[]
  rule_tags?: string[]
  emotion_tags?: string[]
  execution_note?: string | null
  improvement_note?: string | null
  lesson?: string | null
  screenshot_path?: string | null
  screenshot_paths?: string[]
  review_score?: number | null
  ai_summary?: string | null
}

export interface ReviewStats {
  total_reviews: number
  mistake_tags: Array<{ name: string; count: number }>
  rule_effectiveness: Array<{ name: string; count: number; net_pnl: number; win_rate: number }>
  market_phase: Array<{ name: string; count: number; net_pnl: number; win_rate: number }>
  system_compliance: Array<{ name: string; count: number; net_pnl: number }>
}

export interface ReviewCanonicalLineage {
  schema_version: 'review_canonical_lineage_v1'
  source_type: string
  source_id: number
  strategy_version?: string | null
  input_digest: string
  dataset_keys: CanonicalDatasetKey[]
  manifest_digests: string[]
  window: { start: string; end: string }
  source_window: { start: string | null; end: string | null }
  input_identity: CanonicalInputIdentity
  auxiliary_input_identities?: Record<string, CanonicalInputIdentity>
}

export interface ReviewObservationLineage {
  schema_version: 'review_source_lineage_v1'
  source_type: string
  source_id: number
  source_snapshot_schema_version?: string | null
  source_mode?: string | null
  bar?: {
    bar_start?: string | null
    bar_end?: string | null
    confirmation_mode?: string | null
  } | null
}

export type ReviewFormalLineage = ReviewCanonicalLineage | ReviewObservationLineage

export interface ReviewBarsResponse {
  lineage: ReviewFormalLineage
  bars: BarData[]
}
